lib/rust/ast/generation/src/ast.rs

```diff
@@ -0,0 +1,268 @@
+//! This module exports an internal AST of Rust source code.
+
+use itertools::Itertools;
+use quote::quote;
+use std::ops::Add;
+use syn;
+use syn::Ident;
+use syn::ItemMod;
+
+
+
+// =================
+// === Scala AST ===
+// =================
+
+/// Represents a single source file `$name.scala`.
+#[allow(missing_docs)]
+#[derive(Debug,Clone)]
+pub struct File {
+    pub lib     : Stdlib,
+    pub package : String,
+    pub name    : Name,
+    pub content : syn::File,
+}
+
+impl File {
+    /// Creates a Scala source file.
+    ///
+    /// Note that the content of the file is going to be wrapped in an extra object
+    /// `object FileName { content.. }`, because Scala (compared to Rust) doesn't support
+    /// top level type aliases.
+    pub fn new(name:&str, package:&str, content:syn::File) -> Self {
+        Self{lib:Stdlib(), package:package.into(), name:name.into(), content}
+    }
+}
+
+/// Represents a modified standard library.
+#[derive(Debug,Clone,Copy,Default)]
+pub struct Stdlib();
+
+/// Represents an object `object Name { line; ... }`.
+#[allow(missing_docs)]
+#[derive(Debug,Clone)]
+pub struct Module<'a> {
+    pub name  : Name,
+    pub lines : &'a [syn::Item],
+}
+
+/// Represents a class `class Name[Generics](field:Field, ...)`.
+///
+/// This can either correspond to a Rust struct or a single enum variant.
+#[allow(missing_docs)]
+#[derive(Debug,Clone,Ord,PartialOrd,Eq,PartialEq)]
+pub struct Class {
+    pub typ   : Type,
+    pub args  : Vec<(Name,Type)>,
+    /// True iff the class fields have user defined names.
```

Comment:
If you have a doc you should describe what it is, not just the conditions for setting it.